### PR TITLE
Implement map and revisit write to hocon functionality + incremental code clean up

### PR DESCRIPTION
**Pull request by afsalthaj**

### CODE REVIEW

File: core/src/main/scala/zio/config/ReadError.scala [hunk 1]
```diff
@@ -9,6 +10,24 @@ sealed trait ReadError[A] extends Exception { self =>
...
+    }
+
+  def atIndex(index: Int): ReadError[A] =
+    self match {
+      case ReadError.MissingValue(path)             => ReadError.MissingValue(path :+ Step.Index(index))
+      case ReadError.FormatError(path, message)     => ReadError.FormatError(path :+ Step.Index(index), message)
+      case ReadError.ConversionError(path, message) => ReadError.ConversionError(path :+ Step.Index(index), message)
+      case ReadError.OrErrors(list)                 => ReadError.OrErrors(list.map(_.atIndex(index)))
+      case ReadError.AndErrors(list)                => ReadError.AndErrors(list.map(_.atIndex(index)))
+    }
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ReadError.scala:29]: There is a lot more to be done in this side to avoid redundant errors, and proper indexing.
A part of it is done in this PR. We will continue the rest in another PR (as this one is majorly map implementation and cleaning up write-to-hocon functionality)

File: core/src/main/scala/zio/config/ConfigDocs.scala [hunk 2]
```diff
@@ -12,4 +12,5 @@ object ConfigDocs {
   final case class Both[K, V](left: ConfigDocs[K, V], right: ConfigDocs[K, V])  extends ConfigDocs[K, V]
   final case class OneOf[K, V](left: ConfigDocs[K, V], right: ConfigDocs[K, V]) extends ConfigDocs[K, V]
   final case class Sequence[K, V](element: List[ConfigDocs[K, V]])              extends ConfigDocs[K, V]
+  final case class Map[K, V](path: K, element: ConfigDocs[K, V])                extends ConfigDocs[K, V]
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ConfigDocs.scala:15]: https://github.com/zio/zio-config/issues/282

These things only in follow up PR

File: core/src/main/scala/zio/config/ConfigDocsFunctions.scala [hunk 3]
```diff
@@ -62,32 +71,54 @@ private[config] trait ConfigDocsFunctions {
...
+                case _ => Map(key, loop(tree, flattened, docs, keys))
+              }
 
             case Sequence(element :: Nil) =>
-              tree.getPath(initialValue.toList) match {
+              tree.getPath(keys.toList) match {
                 case PropertyTree.Sequence(value) if value.nonEmpty =>
                   Sequence(value.map(t => loop(t, t.flatten, element, Vector.empty)))
-                case _ => Sequence(loop(tree, flattened, element, initialValue) :: Nil)
+                case _ => Sequence(loop(tree, flattened, element, keys) :: Nil)
```
Comment (afsalthaj) [inline core/src/main/scala/zio/config/ConfigDocsFunctions.scala:121]: ConfigDocsFunctions to be revisited in another PR

File: core/src/test/scala/zio/config/ListsCornerCasesTest.scala [hunk 4]
```diff
@@ -131,17 +196,24 @@ object ListsCornerCasesTest
...
 
-          val cCfg = (string("a") |@| list("b")(listStrict(string("c"))))(Cfg, Cfg.unapply)
+          val cCfg = (string("a") |@| list("b")(listStrict(string("c"))))(
+            Cfg,
+            Cfg.unapply
+          )
 
           val res = read(
-            cCfg from ConfigSource(
+            cCfg from ConfigSource.mk(
```
Comment (afsalthaj) [inline core/src/test/scala/zio/config/ListsCornerCasesTest.scala:205]: Remove ConfigSource.mk

File: core/src/main/scala/zio/config/ConfigDocs.scala (hunk 2 above)
Comment (jdegoes) [reply core/src/main/scala/zio/config/ConfigDocs.scala:15]: Shouldn't this be, `DynamicMap(value: Map[K, ConfigDocs[K, V]])`?

File: core/src/main/scala/zio/config/ConfigSource.scala [hunk 5]
```diff
@@ -16,38 +16,30 @@ import zio.Task
...
   private[config] val SystemEnvironment    = "system environment"
   private[config] val SystemProperties     = "system properties"
   private[config] val CommandLineArguments = "command line arguments"
 
-  def empty[K, V]: ConfigSource[K, V]                                       = TreeConfigSource(PropertyTree.empty, Set.empty)
-  def apply[K, V](tree: PropertyTree[K, V], sourceDescription: Set[String]) = TreeConfigSource(tree, sourceDescription)
+  def empty[K, V]: ConfigSource[K, V] =
+    ConfigSource(Set.empty, _ => PropertyTree.empty)
+
+  def mk[K, V](tree: PropertyTree[K, V], descriptions: Set[String]): ConfigSource[K, V] =
```
Comment (jdegoes) [inline core/src/main/scala/zio/config/ConfigSource.scala:41]: This should probably be renamed, `fromPropertyTree` and sorted alphabetically.

File: core/src/main/scala/zio/config/ReadFunctions.scala [hunk 6]
```diff
@@ -77,15 +85,37 @@ private[config] trait ReadFunctions {
...
+      cfg.source.getConfigValue(keys.reverse) match {
+        case PropertyTree.Leaf(_)     => formatError(path, "Leaf", "Record")
+        case PropertyTree.Sequence(_) => formatError(path, "Sequence", "Record")
+        case PropertyTree.Record(values) =>
+          val result: List[(K, Res[B])] = values.toList.zipWithIndex.map {
+            case ((k, tree), idx) =>
+              val source: ConfigSource[K, V] =
+                ConfigSource(cfg.source.sourceDescription, tree.getPath)
+
+              (k, loopAny(Step.Index(idx) :: path, Nil, cfg.config.updateSource(_ => source)))
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadFunctions.scala:111]: `Step.Index` is for `Sequence` only. `Key("a"), Key("b"),Index(5).Key("c")` means `a.b[5].c`.
In `Map::toList` index is unstable (so user can't associate index in error message with exact value) and redundant - path in map bases on `Key`s only is enough to find exact value.

File: core/src/main/scala/zio/config/ReadFunctions.scala [hunk 7]
```diff
@@ -77,15 +85,37 @@ private[config] trait ReadFunctions {
...
+          }
+
+          seqMap2[K, ReadError[K], B, ReadError[K]]((index, k, error) => error.atKey(k).atIndex(index))(result.toMap).swap
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadFunctions.scala:114]: You don't need `atKey` and `atIndex` here (or anywhere else) since we have full path in `path` parameter.
Please, add test for error path if you think it's necessary.

File: core/src/main/scala/zio/config/ConfigSource.scala (hunk 5 above)
Comment (afsalthaj) [reply core/src/main/scala/zio/config/ConfigSource.scala]: There is already `fromPropertyTree`. this was a carry-over from previous merges. I have removed the function as such and reused `fromPropertyTree`

File: core/src/main/scala/zio/config/ReadFunctions.scala (hunk 7 above)
Comment (afsalthaj) [reply core/src/main/scala/zio/config/ReadFunctions.scala:114]: This is already in another branch as a follow up PR

File: core/src/main/scala/zio/config/ReadFunctions.scala [hunk 8]
```diff
@@ -77,15 +85,37 @@ private[config] trait ReadFunctions {
...
+        case Right(a) =>
+          cfg.f(a).swap.map(ReadError.ConversionError(path.reverse, _)).swap
+      }
+
+    def loopMap[B](path: List[Step[K]], keys: List[K], cfg: DynamicMap[K, V, B]): Res[Map[K, B]] =
+      cfg.source.getConfigValue(keys.reverse) match {
+        case PropertyTree.Leaf(_)     => formatError(path, "Leaf", "Record")
+        case PropertyTree.Sequence(_) => formatError(path, "Sequence", "Record")
+        case PropertyTree.Record(values) =>
+          val result: List[(K, Res[B])] = values.toList.zipWithIndex.map {
```
Comment (senia-psm) [inline core/src/main/scala/zio/config/ReadFunctions.scala:106]: You don't `zipWithIndex` here.
